### ISSUE TITLE
Improve the operator proposal for supporting providers deploying additional controllers via the component.yaml

### ISSUE THREAD

**Issue by asalkeld** [kubernetes-sigs/cluster-api#4982]
I am trying to implement the API and something does not make sense..

For example: see https://github.com/metal3-io/cluster-api-provider-metal3/releases/tag/v0.4.1
there are 5 deployments
```
 name: capm3-controller-manager
 namespace: capi-webhook-system

 name: capm3-ipam-controller-manager
 namespace: capi-webhook-system

 name: capm3-controller-manager
 namespace: capm3-system

 name: capm3-ipam-controller-manager
 namespace: capm3-system

 name: capm3-baremetal-operator-controller-manager
 namespace: capm3-system
```
Note: most of these have a container with the name "manager" (kubebuilder default) so we can't rely on the container name.

Problem 1:
there is no name in the DeploymentSpec/MangerSpec

Problem 2:
there is only one DeploymentSpec/Manager per provider, should it be a map/list?

Problem 3:
With the move to placing everything in one "targetNamespace", two of the above deployments will be overwritten as they will have
the same name when this runs https://github.com/kubernetes-sigs/cluster-api/blob/master/cmd/clusterctl/client/repository/components.go#L305

/kind bug
/area operator
 
/cc @fabriziopandini

**Comment by asalkeld** [comment]
Note for Problem 3, I am working on https://github.com/asalkeld/cluster-api/commit/95a3a2f61aa2b96efbbd56402e1b79e970a4416f

**Comment by fabriziopandini** [comment]
If I got it right, capm3 is installing some other controller in addition to the provider controller using the same component.yaml, and from your issue, I assume your expectation is to pilot those components via the operator API.

But clearly, as you are pointing out this is something not covered as a use case in the current proposal/API design the community agreed upon.

I defer to the usual discussion process if this use case is something we should support or not and how to generalise this use case so all the providers can benefit from it, but most probably this will require an amendment to the proposal, to the API types, and most likely to the clusterctl contract as well.

As a consequence, my very first reaction is that its implementation should be deferred to the second iteration of the operator work, mostly because I really would like to avoid further delay on the ongoing activities and there is still much ground to be covered before completing what currently in scope.

/remove-kind bug
/kind feature

/retitle Improve the operator proposal for supporting providers deploying additional controllers via the component.yaml

**Comment by asalkeld** [comment]
@fabriziopandini thanks for the response.

First a clarification: 
It seems like most providers have 1 main controller (metal2 has 2) that runs that is a deployment by default in cap?-system namespace AND a webhook that runs in a different deployment in capi-webhook-system namespace. 
Which deployment should we alter if there is a ProviderSpec.DeploymentSpec? Both? This would make sense for image but not really for args..

we can separate this into 2 issues:

1. still assume that we will only modify one deployment, that is fine as a first step, but still how do we identify which deployment to modify? What I suggest is the DeploymentSpec and MangerSpec have a "name" or "selector" field so we can isolate which deployment to modify.
2. phase 2, we turn this into a list/map so we can modify multiple deployments (metal3 has 2 main controllers and 2 webhooks).

**Comment by sbueringer** [comment]
Regarding Problem 3

@asalkeld Not sure if it helps, and I don't have enough context / want to interfere. With CAPI v1alpha3 we deployed all controllers twice (once for the reconciler, once for the webhooks). Starting with v1alpha4 we dropped that and all other providers are running just once (with reconciler **and** webhooks). (ref: https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/book/src/developer/providers/v1alpha3-to-v1alpha4.md#required-kustomize-changes-to-have-a-single-manager-watching-all-namespaces-and-answer-to-webhook-calls)

So starting with v1alpha4 those should only be 3 deployments and the two from capi-webhook-system shouldn't exist anymore. Not sure if that helps / makes sense :)
